src/main/java/com/alvarium/annotators/ScaAnnotator.java

```diff
@@ -72,7 +78,24 @@ private Map<String, String> getPackages(String dir) throws AnnotatorException {
         } catch (Exception e) {
             throw new AnnotatorException("Failed to get dependency info", e);
         }
-    }   
+    }
+
+    private List<String> retrievePackagesVulnerabilities( Map<String, String> packages) throws AnnotatorException {
+        try{
+            final VulnerabilityApiHandler apiHandler = new VulnerabilityApiHandlerFactory().getHandler("osv"); //To replace with a var from config
+            JSONObject requestBody = apiHandler.buildQueryRequestBody(packages);
+            JSONObject response = apiHandler.makeApiCall(requestBody);
+            List<String> vulnerabilities = apiHandler.parseResponse(response);
```

Comment:
I think it can be done in `fetchVulnerabilityIds ` without issues.
But what is the benefit of splitting into 2 new classes?

Reply:
Both are fine approaches, I prefer typing requests and responses because they act as contracts between the client and the server as well as:
1. You can easily know what the request and the response contain without looking at the API documentation and just by looking at the request class (intellisense is a plus)
2. If you are going to use the request in another API call/method, its always easier to have the logic in one place (you already have it in the `buildQueryRequestBody()` method which is good, it's basically acting as a class at this point but we also mentioned exposing only one method so this one won't be accessible) 
3. It provides a decent place for serialization/deserialization logic to live in

See this code snippet below for what I had in mind 

```java
class OsvBatchRequest {
    public List<Query> queries;

    OsvBatchRequest(Map<String,String> packages) {
        queries = new List<Query>();
        for (Map.Entry<String, String> entry : packages.entrySet()) {
            Package package = new Package(entry.getKey());
            queries.add(
                new Query(entry.getValue(), pacakge);
            )

        }
    }

    static class Query {
        public Package package;
        public String version;
        
        Query(String version, Package package) {
            this.version = version;
            this.package = package;
        }
    }

    static class Package {
        public String name;
        public String ecosystem;

        Package(String name) {
            this.name = name;
            this.ecosystem= 'maven';
        }
    }
    
}
```

```java
OsvBatchRequest req = new OsvBatchRequest(packages);
StringEntity entity = new StringEntity(gson.toJson(req), StandardCharsets.UTF_8);
httpPost.setEntity(entity);
```

Reply:
I followed the structure you suggested. The only issue I have now is the apiHandler var name and the Interface name too.
Following the implementation now, we're interactiong with the OsvBatchResponse from the annotator. Adding to that the BatchResponse implements the VulnerabilityApiHandler interface.
I am just not sure about the Interface naming + should the annotator interact with the BatchRequest directly or if there should be any middleware.

Reply:
Also another less important design question is if the packages should be passed directly in the `OsvBatchResponse` constructor.

Therefore executing these lines
`
OsvBatchRequest req = new OsvBatchRequest(packages);
`
`
StringEntity entity = new StringEntity(gson.toJson(req), StandardCharsets.UTF_8);
`
in the constructor instead of when executing the `makeApiCall` method.

Reply:
Sorry I missed these replies while reviewing. I think the first comment might have been answered in the review comments. As for the second question, I think it's fine either way, I'd prefer what you have already implemented since it doesnt expose the request where it isnt needed.